base/loading.jl

```diff
@@ -419,42 +744,50 @@ function _require(mod::Symbol)
         if doneprecompile === true || JLOptions().incremental != 0
             # spawn off a new incremental pre-compile task for recursive `require` calls
             # or if the require search declared it was pre-compiled before (and therefore is expected to still be pre-compilable)
-            cachefile = compilecache(mod)
-            m = _require_from_serialized(mod, cachefile)
+            cachefile = compilecache(into, mod)
+            m = _require_from_serialized(into, mod, cachefile)
             if isa(m, Exception)
                 @warn "The call to compilecache failed to create a usable precompiled cache file for module $name" exception=m
                 # fall-through, TODO: disable __precompile__(true) error so that the normal include will succeed
             else
-                return
+                return info
             end
         end
 
         # just load the file normally via include
         # for unknown dependencies
+        if info isa Tuple{String,String,UUID}
+            # horrible hack, but since we don't control module creation, ¯\_(ツ)_/¯
+            ccall(:jl_set_global, Cvoid, (Any, Any, Any), __toplevel__, ENVINFO, info[2:end])
+        end
```

Comment:
If we constructed the module object to be loaded and then evaluated code into it, then we could explicitly set the `ENVINFO` name in there instead of doing nasty stuff like this. I think we should probably do that, but I didn't want to couple this change with a change like that, so I just hacked around it instead.